### ISSUE TITLE
Preserve line breaks in scoped code

### ISSUE THREAD

**Issue by hirasso** [humbug/php-scoper#1083]
## Feature Request

Currently, quite some line breaks are being stripped from scoped code. This makes that code hard to read (I'm also scoping my own package's main code).

If it's technically possible, it would be amazing to preserve all line breaks / leave them untouched.

**Comment by theofidry** [comment]
I agree it would be best to preserve formatting if possible. I tried it at some point but had to give up as it requires more work, as PHP-Scoper modifies/inserts some statements.

There was an attempt done at https://github.com/humbug/php-scoper/pull/1051. I think https://github.com/nikic/PHP-Parser/issues/1007#issuecomment-2169722319 opened as part of that PR https://github.com/nikic/PHP-Parser/issues/1007#issuecomment-2169722319

**Comment by hirasso** [comment]
If you ever solve this it would be amazing! I'm sure though you have enough on your plate 🥤

**Comment by hirasso** [comment]
Seems like others are [struggling with the same problems](https://github.com/rectorphp/rector?tab=readme-ov-file#known-drawbacks).

> Rector uses [nikic/php-parser](https://github.com/nikic/PHP-Parser/), built on technology called an abstract syntax tree (AST). An AST doesn't know about spaces and when written to a file it produces poorly formatted code in both PHP and docblock annotations.

**Comment by theofidry** [comment]
The problem is not so much the nature of AST: AST works fine because the formatting information is preserved as part of the nodes data. What is less clear and harder is knowing how to modify that formatting when inserting or removing nodes or alternatively how to handle incoherent token positions within the printer.

There is a native printer that somewhat handles this, but it's not clear to me if I'm missing something within PHP-Scoper or if there is more to address in the PHP-Parser printer

**Comment by kkmuffme** [comment]
I remember rector had the same issue previously, but something in nikic/php-parser was fixed, which means it now generally works for them - see e.g. https://getrector.com/demo/26695f7d-9e51-47a6-882e-67771da4e27a - they modify it, but the lines and spacing are preserved (in unmodified lines)

I think it's worth asking them about it, perhaps it's actually a simple fix

**Comment by theofidry** [comment]
>And what was the reason you didn't merge it back then?

The entire CI was red and some changes were not valid (e.g. `<?php` was being removed).

>Could you perhaps retrigger the CI on that commit to see if it's maybe worth merging it now?

I did so, but I can't really check the results atm. At a glance quite a few seems better, but a few e2e tests are failing. So needs more checking.

>I think it's easier to fork it, remove the rector specificstuff that is irrelevant for us anyway and go with that.

The issue with that is mostly keeping it up to date. But if it works, it's worth asking if they are open moving it to a package.
